include/ceed/ceed.h

```diff
@@ -494,22 +494,36 @@ CEED_EXTERN const char *const CeedQuadModes[];
 /// @ingroup CeedBasis
 typedef enum {
   /// Line
-  CEED_LINE = 1 << 16 | 0,
+  CEED_FE_TOPO_LINE = 1 << 16 | 0,
   /// Triangle - 2D shape
-  CEED_TRIANGLE = 2 << 16 | 1,
+  CEED_FE_TOPO_TRIANGLE = 2 << 16 | 1,
   /// Quadralateral - 2D shape
-  CEED_QUAD = 2 << 16 | 2,
+  CEED_FE_TOPO_QUAD = 2 << 16 | 2,
   /// Tetrahedron - 3D shape
-  CEED_TET = 3 << 16 | 3,
+  CEED_FE_TOPO_TET = 3 << 16 | 3,
   /// Pyramid - 3D shape
-  CEED_PYRAMID = 3 << 16 | 4,
+  CEED_FE_TOPO_PYRAMID = 3 << 16 | 4,
   /// Prism - 3D shape
-  CEED_PRISM = 3 << 16 | 5,
+  CEED_FE_TOPO_PRISM = 3 << 16 | 5,
   /// Hexehedron - 3D shape
-  CEED_HEX = 3 << 16 | 6,
+  CEED_FE_TOPO_HEX = 3 << 16 | 6,
 } CeedElemTopology;
 CEED_EXTERN const char *const CeedElemTopologies[];
 
+/// Type of FE space;
+/// @ingroup CeedBasis
+typedef enum {
+  /// L2 FE space
+  CEED_FE_SPACE_L2 = 0,
+  /// H1 FE space
+  CEED_FE_SPACE_H1 = 1,
+  /// H(div) FE space
+  CEED_FE_SPACE_HDIV = 2,
+  /// H(curl) FE space
+  CEED_FE_SPACE_HCURL = 3,
+} CeedFESpace;
+CEED_EXTERN const char *const CeedFESpaces[];
```

Comment:
I'm also a little concerned about this amount of finite element terminology. Do we need to talk about these spaces or could we just say which subset of {values, gradients, divergence, curl} is supplied (all on a reference element)?

Reply:
Question - can the Hdiv interp and grad be written to run from the same code as the H1 interp and grad? If so, then we can simplify this PR a fair amount. 

In terms of the printing, we could change this just to print whichever of the interp, grad, and div matrices are avaliable

Reply:
In my mixed PR, I used P0, L^2 basis which has only interpolation matrix of size (QxP=1) and it is constant at all quadrature points. I tried to create a constant interpolation using H^1, but got error (P must be >1). 
I am not sure that we can use H^1 interpolation for H(div) case. In H(div), interpolation is a matrix of size `(dim*QxP)`. The interpolation for x-component is different from y-component (or z-comp). But in H^1 we use the same interpolation matrix of size `(QxP)` for all other components.

Reply:
Hmm, they're both (as non-tensor) just matrix multiplies. The shape calculation is different as you say. Though I think components (living in E-vector) can be handled the same way as we do for H^1 (Kronecker with identity).

@rezgarshakeri I think you can call `CeedBasisCreateH1` with `P`=1. You can't call `CeedBasisCreateTensorH1Lagrange` because we're hard-coded to putting nodes at Lobatto points and you can't have just one Lobatto point. We could make a variant that puts dofs at Lobatto points, but then you can't make globally H^1 conforming spaces (without significant pain).

Reply:
Right, I can give a constant interpolation array to the `CeedBasisCreateH1`.
Right now I had to create my `dm` using 1 component and 2 dof on each face to give me exactly linear BDM element. So the dmplex, is another part that we need to handle it.

Reply:
Hmm, the shape being different is an issue. Would it help to let the basis set ncomp_dofs and ncomp_qpts separately? That could either make our  ref backend code smaller and more flexible or just more confusing.

Reply:
What do you mean by basis set ncomp_dofs and ncomp_qpts separately?

Reply:
I'm imagining for a 3D two component H1 element ncomp_nodes = ncomp_qpts = 3 but for a 3D two component Hdiv element ncomp_nodes = 2 and ncomp_qpts = 6

Reply:
I have created/used `CeedBasisCreateL2` in my mixed branch (not include it in this PR to make review easier). Do you want me delete that and use `CeedBasisCreateH1`?

Reply:
> Question - can the Hdiv interp and grad be written to run from the same code as the H1 interp and grad? If so, then we can simplify this PR a fair amount.
> 
> In terms of the printing, we could change this just to print whichever of the interp, grad, and div matrices are avaliable

For the GPU backends this will require some refactoring and redesign to take into account that the number of dofs per dimension is different for each component on tensor elements.

Reply:
Right, I was thinking the GPU work would come in a follow up PR and right now we'd merge with the standard "not implemented" message.

Reply:
Regarding the initialization of tensor basis for H(div) and H(curl) (which require two different "H1 1D basis" -- GL and GLL), it may be interesting to spend a little extra time on the design to make it generic enough to allow spectral finite element methods too.

Reply:
Do you have a favorite reference on tensor decomposition for H(div) and H(curl) spaces? We have an immediate need for the non-tensor case, but I agree that libCEED will want high order spaces at some point. I assumed it would be a specialized constructor for tensor H(div) bases.

Reply:
[This book](https://link.springer.com/book/10.1007/978-94-017-7761-2) describes the tensor decomposition p.86

Reply:
Thanks for sharing.

Reply:
I need to digest that case first and extend it to BDM element case we have here.